Gotify individual application support in the RequestURI.

IncognitoCoding opened the pull request:
The current bridge only allows one statically set Gotify URL receiver in the alertmanager.yaml file. This addition would allow users to add a token to the RequestURI that would override the default application token. It would be beneficial to support multiple Gotify applications in the bridge to break out alerts based on application. This would allow multiple receivers in alertmanager.yaml file for different applications.

The change would not affect any existing functionality, so users would not need to modify any current configuration to use this additional functionality.

### Usage Examples
**_Dedicated application token:_**
- Overrides the default token with the dedicated application token.
```shell
curl http://127.0.0.1:8080/gotify_webhook?token=PG3sgy-fWs.jW-gE. -d '
{ "alerts": [
  {
    "annotations": {
      "description":"A description",
      "summary":"A summary",
      "priority":"critical"
    },
    "status": "firing",
    "generatorURL": "http://foobar.com"
  }
]}
'
```
**_Dedicated application with missing token:_**
- Uses the default token.
```shell
curl http://127.0.0.1:8080/gotify_webhook?token= -d '
{ "alerts": [
  {
    "annotations": {
      "description":"A description",
      "summary":"A summary",
      "priority":"critical"
    },
    "status": "firing",
    "generatorURL": "http://foobar.com"
  }
]}
'
```
**_Default application token:_**
- Uses the default token.
```shell
curl http://127.0.0.1:8080/gotify_webhook -d '
{ "alerts": [
  {
    "annotations": {
      "description":"A description",
      "summary":"A summary",
      "priority":"critical"
    },
    "status": "firing",
    "generatorURL": "http://foobar.com"
  }
]}
'
```

IncognitoCoding committed:
Gotify individual application support in the RequestURI.

IncognitoCoding committed:
Added Go templating function processing, such as reReplaceAll, match, and humanize.

Feature: Added Go templating function processing, such as reReplaceAll, match, and humanize.
Update: Because the base of the templating is done through an existing Prometheus repository, the go.mod and go.sum files have been re-created. This includes running the "go mod tidy".

A large chunk of the functions listed under Grafana functions (https://grafana.com/docs/grafana/latest/alerting/fundamentals/annotation-label/template-functions) is usable with the bridge.

From what I could tell, some template functions need more data, which does not pass in the JSON. These functions will log output explaining the function is not supported by the bridge. The functionality for these excluded functions still exists and would be simple to enable if the data becomes avaliable. Also, a few of the template functions in the template.go file were not listed in the Grafana documentation, so those are also excluded.

IncognitoCoding commented:
### **Additional Functionality Added**

Grafana templating offers the ability to reReplaceAll or match, which I found helpful when breaking logs out into smaller chunks. I figured this would be a valuable option on the bridge, so I implemented this functionality.

A good chunk of the functions listed in the Grafana Template functions documentation ([Grafana Templates](https://grafana.com/docs/grafana/latest/alerting/fundamentals/annotation-label/template-functions/)) work with the bridge. A few are not functioning, and I created a log to alert the user. 


Grafana Usage Example:
```shell
 {{ reReplaceAll ".+\\|" " " .Labels.log }}
```

Curl Testing Examples:
```
curl http://127.0.0.1:8080/gotify_webhook -d '
{ "alerts": [
  {
    "annotations": {
      "description": "{{ match \"my+\" \"my|test\" }}",
      "summary":"A summary",
      "priority":"critical"
    },
    "status": "firing",
    "generatorURL": "http://foobar.com"
  }
]}
'
```

```shell
curl http://127.0.0.1:8080/gotify_webhook?token=16gdfass-gE. -d '
{ "alerts": [
  {
    "annotations": {
      "description": "{{ humanize 1234567.0 }}",
      "summary":"A summary",
      "priority":"critical"
    },
    "status": "firing",
    "generatorURL": "http://foobar.com"
  }
]}
'
```

```shell
curl http://127.0.0.1:8080/gotify_webhook -d '
{ "alerts": [
  {
    "annotations": {
      "description": "{{ reReplaceAll \"fir\" \" \" .Status }}",
      "summary":"A summary",
      "priority":"critical"
    },
    "status": "firing",
    "generatorURL": "http://foobar.com"
  }
]}
'
```

DRuggeri commented on go.mod:
```diff
@@ -3,8 +3,9 @@ module github.com/DRuggeri/alertmanager_gotify_bridge
-	github.com/prometheus/client_golang v1.10.0
-	github.com/prometheus/common v0.20.0
+	github.com/dustin/go-humanize v1.0.1
+	github.com/prometheus/client_golang v1.14.0
```
Thanks for bumping these up while in the neighborhood!

DRuggeri commented on main.go:
```diff
@@ -202,11 +203,33 @@ func main() {
 	metrics["requests_received"]++
 
+	if strings.Contains(r.RequestURI, "/gotify_webhook?token=") {
+		splitToken := strings.Split(r.RequestURI, "/gotify_webhook?token=")
```
This feels a bit too rigid (maybe we add other params later) - can this please be adjusted to use the following instead?
```
import (
...
  "net/http"
...
)
...

  appToken := r.URL.Query().Get("token")
```
This would also remove one of the if/else statements

In fact, using this approach, I like the idea that practically ANY parameter could be overridden - so great idea!

IncognitoCoding replied:
Nice enhancement! I made this adjustment.

DRuggeri commented on main.go:
```diff
@@ -441,18 +464,48 @@ func (svr *bridge) handleCall(w http.ResponseWriter, r *http.Request) {
-		return "", fmt.Errorf("error in Template: %s", err)
-	}
+	if unsupportedFunc == "" {
+		titleTemplate := template.NewTemplateExpander(context.Background(), templateString, "tmp", data, 0, nil, nil, nil)
```
This is cool - I didn't realize Prometheus had this functionality.

DRuggeri commented on main.go:
```diff
@@ -441,18 +464,48 @@ func (svr *bridge) handleCall(w http.ResponseWriter, r *http.Request) {
+		}
 
-	return templatedTitle.String(), nil
+		result = strings.ReplaceAll(result, "<no value>", "[no value]")
```
Looking through the source code for [template](https://github.com/prometheus/prometheus/blob/main/template/template.go#L143-L145), it looks like this shouldn't be needed since the default is to set `missingkey=zero` - which should avoid the injection of `<no value>`. Have you observed `<no value>` showing up in output?

IncognitoCoding replied:
Thanks for catching this. This line is not required. I forgot to remove it when I was reverse-engineering some code.

DRuggeri commented on main.go:
```diff
@@ -441,18 +464,48 @@ func (svr *bridge) handleCall(w http.ResponseWriter, r *http.Request) {
+	var err error
+	var unsupportedFunc string
+
+	//Excludes unsupported template function calls.
```
Since I'm new to these prometheus templating functions, I'm not sure why these aren't supported. Can you add a little bit of color in the comment or README maybe?

IncognitoCoding replied:
After a good night's rest and looking at the code again, I was able to make the tweaks to allow all the Grafana Template functions to work with the bridge. I overlooked the externalURL being passed outside the "Alerts" JSON section. There are still a few functions excluded, and these are not part of the Grafana Template Functions, so I kept them out of the bridge options for consistency with Grafana.

DRuggeri replied:
Thanks - I mean to ask "why aren't these functions supportable by the bridge?" I am guessing maybe they rely on some other data that is not available? At quick glance, it looks like first, label, value, strValue relies on a `sample` pointer being passed in (which... who knows where we could get that, so perhaps it shouldn't work) while query and safeHtml can accept a string value that may be doable just fine in a template?

In any event, I'm wondering why these functions are singled out. I ask because it may be possible to just ignore if they are unsupported or not by setting the right usage expectations in documentation.

DRuggeri reviewed (changes requested):
I really like the idea of a query string parameter override allowing the user to choose the application token as well as expanding the templating capabilities to use Prometheus templates.

In addition to the small tweaks noted during review, can you please also:
* Update README (and maybe even usage) to show that the token can be overridden
* Update README to note the expanded templating support and what functions are available (a link to the prometheus function list is fine, but just calling out which ones won't work)

IncognitoCoding committed:
Optimized the token override and allowed all Template functions.

Enhancement: Converted the override token catch to a more efficient configuration.
Functionality: Added the externalURL to allow all Grafana template functions to work with the bridge.
ReadMe: Revised the ReadMe file to explain the override and Grafana template function usage.

IncognitoCoding commented:
All changes have been made, including the ReadMe. I was thinking about initially doing the ReadMe file, but I wanted to check that you liked the updates before making those edits.

I just set up Grafana last week for additional monitoring, and I have been using Gotify for other notifications for a while now. Grafana's webhook can send alerts to Gotify (with no priority and some strange formatting), but I realized the limitation with Grafana alerts is it can not alert on 'Mimir / Cortex / Loki' even though the dashboard is aware of the firing alerts. This incompatibility had me searching for an option when I came across your bridge. Since I could do separate contacts in Grafana for Gotify and custom templates, I wanted to work to integrate these into your bridge. I was just aware of the Prometheus function templates yesterday. It took some searching, but I was happy to find they had them created, so no wheel needed to be reinvented. The most challenging part was I primarily coded in Python and never wrote in Go before, so I had to bring myself up to speed on Go language to request these changes.

Excellent work on creating this bridge, and thank you for allowing the integrations so I did not have to maintain a separate repository. These changes will enable me to have one centralized alert distribution point with Gotify application routing for everything.

DRuggeri commented on main.go:
```diff
@@ -202,11 +205,24 @@ func main() {
 
+	appToken := r.URL.Query().Get("token")
+	if appToken != "" {
+		fmt.Printf("Gotify application token (%s) found in request URI - overriding default token: (%s)\n", appToken, *svr.gotifyToken)
```
Sorry for missing this in the previous review - can you please tweak this fmt.Printf to be log.Debugf?

IncognitoCoding replied:
I didn't even notice I had the wrong output selection. I will revise.

DRuggeri commented on main.go:
```diff
@@ -255,6 +271,13 @@ func (svr *bridge) handleCall(w http.ResponseWriter, r *http.Request) {
+			if alert.ExternalURL != "" {
+				externalURL, err = url.Parse(alert.ExternalURL)
+				if err != nil {
+					fmt.Printf("External URL Format Error: %s", err)
```
Can you please also tweak this one to be log.Printf for consistent formatting?

IncognitoCoding replied:
I didn't even notice I had the wrong output selection. I will revise.

IncognitoCoding committed:
Token override message wrapped with the debug flag.

Revision: The token override debug message was not re-implemented after the token override enhancement. Added the debug wrapper to the message.

DRuggeri commented on main.go:
```diff
@@ -48,6 +49,7 @@ type Alert struct {
 	GeneratorURL string
 	StartsAt     string
 	ValueString  string
+	ExternalURL  string
```
I think this maybe should this be GeneratorURL? I don't see a reference to ExternalURL being used in the [alert object](https://github.com/prometheus/alertmanager/blob/v0.25.0/template/template.go#L251) that alertmanager will send.

IncognitoCoding replied:
Honestly, I did not dig into them that deep because I wanted only to include functions that were an option in [Grafana Template functions](https://grafana.com/docs/grafana/latest/alerting/fundamentals/annotation-label/template-functions) These being excluded were additional functions that Prometheus had listed. Grafana did not list them as an option. I figured for consistency purposes, excluding them would be best to keep the template functions that the bridge supports the same as Grafana. Plus, it simplifies our documentation in the ReadMe. Thoughts?

DRuggeri replied:
I just replied in the other thread, but I think we could just drop this and state that we're using the [Prometheus template functions ](https://prometheus.io/docs/prometheus/latest/configuration/template_reference/) since that is what is really happening under the covers. My guess is that the Grafana implementation is doing the same thing

DRuggeri commented on main.go:
```diff
@@ -255,6 +271,13 @@ func (svr *bridge) handleCall(w http.ResponseWriter, r *http.Request) {
 				log.Printf("    Alert %d", idx)
 			}
 
+			if alert.ExternalURL != "" {
```
See prior note - I don't think ExternalURL will ever be sent, but rather GeneratorURL will be. That said, I do feel a little nervous that the prometheus template functions [use this value without checking that it is a non-nil pointer](https://github.com/prometheus/prometheus/blob/main/template/template.go#L342-L344).

Since I think it is possible to receive an alert that does not contain an ExternalURL field, it may not be safe to include this function.

IncognitoCoding replied:
I didn't realize "ExternalURL" was exclusive to Grafana. After further testing, you are correct. It is exclusive to Grafana, but the result of no "ExternalURL" results in the same error when generating the template, so it doesn't cause any unexpected crashes. Still, that output could confuse an end user.
- error in Template: error executing template tmp: template: tmp:1:3: executing "tmp" at <externalURL>: error calling externalURL: runtime error: invalid memory address or nil pointer dereference

We could leave it as a Grafana alert exclusive? See code below.

```go
func renderTemplate(templateString string, data interface{}, externalURL *url.URL) (string, error) {
	var result string
	var err error
	var unsupportedFunc string

	//Excludes unsupported template function calls.
	switch {
	case (strings.Contains(templateString, "{{ query") || strings.Contains(templateString, "{{query")):
		unsupportedFunc = "query"
	case (strings.Contains(templateString, "{{ first") || strings.Contains(templateString, "{{first")):
		unsupportedFunc = "first"
	case (strings.Contains(templateString, "{{ label") || strings.Contains(templateString, "{{label")):
		unsupportedFunc = "label"
	case (strings.Contains(templateString, "{{ value") || strings.Contains(templateString, "{{value")):
		unsupportedFunc = "value"
	case (strings.Contains(templateString, "{{ strvalue") || strings.Contains(templateString, "{{strvalue")):
		unsupportedFunc = "strvalue"
	case (strings.Contains(templateString, "{{ safeHtml") || strings.Contains(templateString, "{{safeHtml")):
		unsupportedFunc = "safeHtml"
	case (strings.Contains(templateString, "{{ sortByLabel") || strings.Contains(templateString, "{{sortByLabel")):
		unsupportedFunc = "sortByLabel"
	case (strings.Contains(templateString, "{{ externalURL") || strings.Contains(templateString, "{{externalURL")) && fmt.Sprint(externalURL) == "<nil>":
		unsupportedFunc = "externalURL"
	case (strings.Contains(templateString, "{{ pathPrefix") || strings.Contains(templateString, "{{pathPrefix")) && fmt.Sprint(externalURL) == "<nil>":
		unsupportedFunc = "pathPrefix"
	default:
		unsupportedFunc = ""
	}

	if unsupportedFunc == "" {
		titleTemplate := template.NewTemplateExpander(context.Background(), templateString, "tmp", data, 0, nil, externalURL, nil)
		result, err = titleTemplate.Expand()
		if err != nil {
			return "", fmt.Errorf("error in Template: %s", err)
		}
		return result, err
	} else if unsupportedFunc == "externalURL" || unsupportedFunc == "pathPrefix" {
		return "", fmt.Errorf("error in Template: The '%s' template function is only supported when the alert is generated through Grafana", unsupportedFunc)
	} else {
		return "", fmt.Errorf("error in Template: The bridge does not support the function %s", unsupportedFunc)
	}
}
```

DRuggeri replied:
Since the module in actual use is the prometheus template library, maybe it's best to just say in the README that the functions available in Prometheus templating are available here as well - and provide a link to the [template.go](https://github.com/prometheus/prometheus/blob/main/template/template.go) file.

That way... 
* In the README, we could say something along the lines of, "Additional Prometheus template functions are available for the Grafana use case
* In this function we can just skip all logic for supported/unsupported functions (since a missing function would just be a regular template error)
* And we can ignore `ExternalURL` since it is Grafana-specific and we've called out that this is the Prometheus module

... said another way, we can just not care about what functions are available or not since the Template expansion should result in an error message indicating the same.

Thanks for testing the dereferenced `nil` failure scenario. This would normally be a runtime panic and crash, but since this is being executed in a Handler, I believe the mux wraps all calls with a `recover` to avoid the whole server going down if a bad request comes in and cannot handle.

IncognitoCoding replied:
I looked a little more at the additional functions used in Prometheus. Some functions like `query` and `first` will not have the ability to pass a queryFunction (see template_test.go), so these function would not work.  Also, the few you already glanced at, like `label`, `value`, etc. rely on sample data, so those would not work either. 

This only leaves a few extra functions in this module that will integrate with the bridge correctly.

It looks like we have two options.

1. Leave the case statements listed, and say the bridge supports the functions listed in the Grafana Template Functions link. No additional explanation is required, and this could limit user confusion.

2. Let all functions pass with a link to the Grafana Template Functions and Template.go file. Some additional information would be required to explain all functions in the Grafana Template Function work. Add another section saying additional functions are available in the Template.go file, but some do not operate in the bridge due to data restrictions. These functions would need to be listed in our ReadMe.

Which option do you like best and would limit any user confusion?

DRuggeri replied:
I think there's confusion about the point I'm making. This PR doesn't support Grafana functions. If you have a look at the `main.go` patch, you will see this import added:
> "github.com/prometheus/prometheus/template"

Then, later in the patch, you are using `template.NewTemplateExpander` - this function call is returning a new [template.Expander struct](https://github.com/prometheus/prometheus/blob/main/template/template.go#L146). This is all coming from https://github.com/prometheus/prometheus/blob/main/template/template.go - so at no point are any of the Grafana packages being used (thus, none of the grafana-only functions being available).

In fact, looking at the Grafana source code, it appears that [it is just using the same prometheus/template package](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/state/template.go#L32-L64) as what your patch is using. This is why I was suggesting just referencing the function names that are in the Prometheus package rather than the Grafana docs - they (indirectly) are the same thing except a few functions added by Grafana.

The way the Grafana application adds custom functions is by [creating a FuncMap](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/state/template_functions.go#L11-L15) and then [adds those functions](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/state/template.go#L58) to the [standard FuncMap in the prometheus/template module](https://github.com/prometheus/prometheus/blob/main/template/template.go)... but since those are all Grafana-specific, it's still best to just point users to the documentation for the [Prometheus template functions](https://prometheus.io/docs/prometheus/latest/configuration/template_reference/) documentation.

IncognitoCoding replied:
There is no confusion on this side regarding how things are called and functioning. I put this together by reading all the Prometheus calls and Grafana to see how they all worked. Initially, I set up a single funcMap with reReplaceAll and got it working, and I realized I should add all the supported functions, which put me down the path of adding in the template.go file.

The reason I put the two options is more for end-users trying to use the bridge and know what and what doesn't work. Since the Go Templating is focused around Grafana in the ReadMe, I was thinking it would be best to show that the Grafana Template functions work alongside the templating without adding any extra overhead confusion around the few additional functions in the template.go file.

I can see where you are coming from in wanting to keep things in-line for reference. I think this falls toward the developer's side the most to explain how things function and what additional features are being used. Let me see what I can put together for documentation in the ReadMe to try and keep the explanation high-level for the general user, but advance comments for anyone doing development or wants to try and use the few additional functions.

DRuggeri reviewed (changes requested):
Thanks for the readme update - looks great!

I'm a little unsure about ExternalURL, though - please see notes

DRuggeri commented:
> The most challenging part was I primarily coded in Python and never wrote in Go before, so I had to bring myself up to speed on Go language to request these changes.

All good - you're doing fine! I did just add a few comments that could possibly lead to a panic (crash in Go-speak) when dealing with ExternalURL. Otherwise, I love the idea!

IncognitoCoding committed:
Code reduction to allow exception handling through the template.go

Reduction: Removed all the unsupportedFunc catches to allow all functions to pass through and the exception handling to be handled through the template.go.
ReadMe Update: Revised the ReadMe template functions to explain in more detail what template functions are supported and how the template functions are generated.

IncognitoCoding commented:
We should be all set. Let me know what you think!
